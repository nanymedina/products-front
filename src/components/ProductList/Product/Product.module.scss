.product {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 4px;
  background-color: #ffffff;
  border: 1px solid #dde3ed;
  border-radius: 4px;
  width: 220px;
}

.product_image {
  width: 100%;
  min-height: 140px;
}

.product_info_content {
  display: flex;
  flex-direction: column;
  padding: 4px;
  min-height: 140px;
}
.product_info {
  min-height: 50px;
  margin-top: 8px;
}

.product_brand {
  margin-right: 5px;
}

.product_description {
  color: #888888;
}

.product_price {
  font-size: 18px;
  color: #615e5e;
  font-weight: bold;
}

.product_discount {
  border-radius: 20px;
  padding: 4px 8px;
  margin-left: 8px;
  background-color: red;
  color: #ffffff;
  font-size: 12px;
  font-weight: bold;
}

.product_old_price {
  text-decoration: line-through;
  color: #888888;
  font-size: 12px;
}
