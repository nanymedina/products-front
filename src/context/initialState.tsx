import StateT from '@src/type/state';

const initialState: StateT = {
  products: [],
  search: null,
  shoppingCart: [],
  error: null,
};

export default initialState;
