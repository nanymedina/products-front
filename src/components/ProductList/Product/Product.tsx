import React, { ReactElement } from 'react';

import ProductT from '@src/type/product';
import Button from '@components/Button';
import { useDispatchApp } from '@src/context/appContext';
import Style from './Product.module.scss';

const Product = (product: ProductT): ReactElement => {
  const { id, brand, description, price, discountRate, oldPrice, image } = product;
  const dispatch = useDispatchApp();

  const handleOnClick = () => {
    dispatch({ type: 'ADD_SHOPPING_CART', params: { product } });
  };

  return (
    <div data-testid="product" className={Style.product}>
      <img className={Style.product_image} src={`svg/${image}`} alt={`product-${id}`} />
      <div className={Style.product_info_content}>
        <div className={Style.product_info}>
          <strong className={Style.product_brand}>{brand}</strong>
          <span className={Style.product_description}>{description}</span>
        </div>
        <div className={Style.product_info_price}>
          <div>
            <span className={Style.product_price}>{`$${price}`}</span>
            {discountRate && <span className={Style.product_discount}>{`${discountRate}%`}</span>}
          </div>
          {oldPrice && <span className={Style.product_old_price}>{`$${oldPrice}`}</span>}
        </div>
      </div>
      <Button data-testid="button" text="Agregar" onClick={handleOnClick} />
    </div>
  );
};

export default Product;
