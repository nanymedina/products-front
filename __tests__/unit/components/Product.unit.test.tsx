import { render, fireEvent } from '@testing-library/react';
import Product from '@components/ProductList/Product';
import { useDispatchApp } from '@src/context/appContext';

jest.mock('@src/context/appContext');

describe('Product', () => {
  const useDispatchAppMock = useDispatchApp as jest.Mock;
  const dispatch = jest.fn();
  useDispatchAppMock.mockReturnValue(dispatch);

  const productWithoutDiscount = {
    id: 123,
    brand: 'brand',
    description: 'description',
    price: 20,
    oldPrice: null,
    discountRate: null,
    image: 'image',
  };

  const productWithDiscount = {
    id: 123,
    brand: 'brand',
    description: 'description',
    price: 10,
    oldPrice: 20,
    discountRate: 50,
    image: 'image',
  };

  afterEach(() => {
    useDispatchAppMock.mockClear();
  });

  it('should show product information without discount', () => {
    const { getByTestId } = render(<Product {...productWithoutDiscount} />);
    const product = getByTestId('product');

    expect(product).toMatchSnapshot();
  });

  it('should show product information with discount', () => {
    const { getByTestId } = render(<Product {...productWithDiscount} />);
    const product = getByTestId('product');

    expect(product).toMatchSnapshot();
  });

  it('dispatch action when click in button', () => {
    const { getByTestId } = render(<Product {...productWithDiscount} />);
    const Button = getByTestId('button');
    const action = { type: 'ADD_SHOPPING_CART', params: { product: productWithDiscount } };

    fireEvent.click(Button);

    expect(dispatch).toBeCalled();
    expect(dispatch).toBeCalledWith(action);
  });
});
