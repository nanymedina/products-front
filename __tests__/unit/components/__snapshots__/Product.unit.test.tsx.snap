// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should show product information with discount 1`] = `
<div
  data-testid="product"
>
  <img
    alt="product-123"
    src="svg/image"
  />
  <div>
    <div>
      <strong>
        brand
      </strong>
      <span>
        description
      </span>
    </div>
    <div>
      <div>
        <span>
          $10
        </span>
        <span>
          50%
        </span>
      </div>
      <span>
        $20
      </span>
    </div>
  </div>
  <button
    data-testid="button"
    type="button"
  >
    <strong>
      Agregar
    </strong>
  </button>
</div>
`;

exports[`Product should show product information without discount 1`] = `
<div
  data-testid="product"
>
  <img
    alt="product-123"
    src="svg/image"
  />
  <div>
    <div>
      <strong>
        brand
      </strong>
      <span>
        description
      </span>
    </div>
    <div>
      <div>
        <span>
          $20
        </span>
      </div>
    </div>
  </div>
  <button
    data-testid="button"
    type="button"
  >
    <strong>
      Agregar
    </strong>
  </button>
</div>
`;
